### PR TITLE
Display the Github status of a release on the overview page

### PR DESCRIPTION
**Pull request by dasch**
Adds an icon indicating the Github Status of each release's commit, as well as a full list of statuses with their descriptions on the release page.

### Tasks
- [ ] :+1: from team
- [ ] Caching and invalidation

### Risks
- Level: Med

### CODE REVIEW

**Comment by dasch** [comment]
@grosser can you take a look at this? Ignore the first commit, that was done in order to be able to run things locally with docker-compose and will be removed before merging.

File: app/views/releases/_release_status.html.erb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<% if status.success? %>
+  <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
```
Comment (ragurney) [inline app/views/releases/_release_status.html.erb]: Could use `icon_tag` helper here as well.

File: app/models/github_status.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  def pending?
+    statuses.any?(&:pending?)
+  end
+
+  def missing?
+    statuses.none?
+  end
+
+  def statuses
```
Comment (ragurney) [inline app/models/github_status.rb]: Maybe cache these on the object, otherwise `pending?`, `missing?` etc. will make this call every time in `release_status.html.erb`.  Something like:
```ruby
def statuses 
  @statuses ||= begin
    ...
  end
end

File: app/models/github_status.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def success?
+    statuses.all?(&:success?)
+  end
+
+  def statuses
+    response = @github.combined_status(@repo, @ref)
+
+    statuses = response.statuses
```
[the remark this answers is not in the document]
Comment (dasch) [reply app/models/github_status.rb]: :+1:

File: app/helpers/releases_helper.rb [hunk 4]
```diff
@@ -8,6 +8,17 @@ def release_label(project, release)
     )
   end
 
+  def status_glyphicon(status_state)
+    icon, text = case status_state
```
[the remark this answers is not in the document]
Comment (dasch) [reply app/helpers/releases_helper.rb]: :+1:

File: app/helpers/releases_helper.rb [hunk 5]
```diff
@@ -8,6 +8,17 @@ def release_label(project, release)
...
+           when "success" then ["ok", "success"]
+           when "failure" then ["remove", "danger"]
+           when "missing" then ["minus", "muted"]
+           when "pending" then ["hourglass", "primary"]
+           end
+
+    %(<span class="glyphicon glyphicon-#{icon} text-#{text}" aria-hidden="true"></span>).html_safe
```
[the remark this answers is not in the document]
Comment (dasch) [reply app/helpers/releases_helper.rb]: The problem is that `icon_tag` doesn't seem to support setting additional classes, since it overwrites the `class` attribute. I frankly don't think it's necessary, since this is already declarative Bootstrap code.

File: app/models/github_status.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    return [] if statuses.nil?
+
+    statuses.group_by {|status| status.context }
```
[the remark this answers is not in the document]
Comment (dasch) [reply app/models/github_status.rb]: Sure, I'll just follow the local practice. I usually use curly brackets when I care about the result rather than the side effect.

File: app/views/releases/_release_status.html.erb (hunk 1 above)
Comment (dasch) [reply app/views/releases/_release_status.html.erb]: If that method was amended to allow passing in additional tags, sure.

File: app/models/github_status.rb (hunk 2 above)
Comment (dasch) [reply app/models/github_status.rb]: :+1:

**Comment by dasch** [comment]
@ragurney care to take another look?

**Comment by dasch** [comment]
I've changed `icon_tag` so that it accepts custom `class` options.

File: .gitignore [hunk 7]
```diff
@@ -46,3 +46,4 @@
 # .bak files
 *.bak
 /vendor/cache/**/
+/db/development.sqlite
```
Comment (grosser) [inline .gitignore]: we already have ` /db/*.sqlite3`

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -146,7 +146,10 @@ def audited_classes
   end
 
   def icon_tag(type, options = {})
-    content_tag :i, '', options.merge(class: "glyphicon glyphicon-#{type}")
+    # Prepend the CSS classes to the existing `css` option if present.
+    css_class = ["glyphicon glyphicon-#{type}", options.fetch(:class, "")].join(" ")
```
Comment (grosser) [inline app/helpers/application_helper.rb]: feels kinda over-engineered / un-precise ...

```Ruby
css_classes = "glyphicon glyphicon-#{type}"
if klass = options[:class]
  css_classes << " #{klass}"
end
```

File: app/models/github_status.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      status_response.statuses.group_by(&:context).map do |context, statuses|
+        Status.new(context, statuses.max_by { |status| status.created_at.to_i })
+      end
+    end
+  end
+
+  private
+
+  def status_response
+    @status_response ||= @github.combined_status(@repo, @ref)
```
Comment (grosser) [inline app/models/github_status.rb]: won't cache on error ... prefer 
```Ruby
return @status_response if defined?(@status_response)
@status_response = 
  begin
    @github.combined_status(@repo, @ref)
  rescue Octokit::Error
    nil
  end

```

File: app/models/github_status.rb [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+  def initialize(repo, ref, github: GITHUB)
+    @github = github
+    @repo = repo
+    @ref = ref
+    @state = nil
+  end
+
+  def state
+    status_response ? status_response.state : "missing"
```
Comment (grosser) [inline app/models/github_status.rb]: here no status_response means it's missing ... and then below it is from empty statuses .. seems kinda confusing

File: app/views/releases/index.html.erb [hunk 11]
```diff
@@ -19,6 +19,7 @@
         <thead>
           <tr>
             <th></th>
+            <th>Status</th>
```
Comment (grosser) [inline app/views/releases/index.html.erb]: unclear that this is github status ...should say github status or have an octopus icon or a hover

File: script/docker_dev_server [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    ./bin/decode_dot_env .env
+fi
+
+if [ -f db/development.sqlite ]; then
+    bundle exec rake db:migrate
+else
+    bundle exec rake db:setup
+fi
+
+exec bundle exec puma -C config/puma.rb
```
Comment (grosser) [inline script/docker_dev_server]: docker things should not be part of this PR

File: app/models/github_status.rb [hunk 13]
```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+class GithubStatus
```
Comment (grosser) [inline app/models/github_status.rb]: we already have commit_status ... this feels like the same code all over again ... can we unify ?

File: .gitignore (hunk 7 above)
Comment (dasch) [reply .gitignore]: Yeah, that's from #2889.

File: app/models/github_status.rb (hunk 9 above)
Comment (dasch) [reply app/models/github_status.rb]: Added memoization of the response on error.

File: app/models/github_status.rb (hunk 10 above)
Comment (dasch) [reply app/models/github_status.rb]: :+1:

File: app/models/github_status.rb (hunk 13 above)
Comment (dasch) [reply app/models/github_status.rb]: That would be a larger refactoring, and I mostly want to fix this UX problem right now...

File: app/views/releases/index.html.erb (hunk 11 above)
Comment (dasch) [reply app/views/releases/index.html.erb]: I was actually considering removing the `th`, since it takes up more space than the icons themselves. How about a hover text saying e.g. "Github status: failed"?

File: script/docker_dev_server (hunk 12 above)
Comment (dasch) [reply script/docker_dev_server]: Will remove after final review, otherwise I can't get the code to run...

File: app/models/github_status.rb (hunk 13 above)
Comment (dasch) [reply app/models/github_status.rb]: It seems that `CommitStatus` is deeply linked to stages...

File: app/models/github_status.rb (hunk 10 above)
Comment (dasch) [reply app/models/github_status.rb]: Changed the code so that we don't try to second-guess the API response's state.

File: app/models/github_status.rb (hunk 13 above)
Comment (grosser) [reply app/models/github_status.rb]: can commit-status use this then ?

File: app/helpers/releases_helper.rb [hunk 14]
```diff
@@ -8,6 +22,14 @@ def release_label(project, release)
     )
   end
 
+  def status_glyphicon(status_state)
```
Comment (grosser) [inline app/helpers/releases_helper.rb]: prefix method+constants with `github_` or `commit_` since we have deploy status which is different

Comment (grosser) [reply app/helpers/releases_helper.rb]: also just `icon` since it does not matter what library we use and the other helper is called `icon_tag`

File: app/models/github_status.rb [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+      status_response.statuses.group_by(&:context).map do |context, statuses|
+        Status.new(context, statuses.max_by { |status| status.created_at.to_i })
+      end
+    end
+  end
+
+  private
+
+  def status_response
+    @status_response ||= begin
```
Comment (grosser) [inline app/models/github_status.rb]: still does not cache on failure since `nil ||=` would execute again

File: app/views/releases/show.html.erb [hunk 16]
```diff
@@ -16,6 +16,13 @@
   <label>Created</label>
   <%= relative_time(@release.created_at) %>
 
+  <legend>Github Status</legend>
```
Comment (grosser) [inline app/views/releases/show.html.erb]: maybe name this commit-status so it's not github specific and can serve multiple backends eventually ...

**Comment by grosser** [comment]
needs some `Rails.cache` love or this will be a neat N+1 on the release page :)
... also make sure we use some sane timeouts since this would block rendering that page

**Comment by dasch** [comment]
@grosser I would assume Octokit has sensible timeouts already?

I wanted to ask about caching – would it be possible to have the Github event webhook invalidate the cache? We'd want short TTLs for new releases but long ones for old releases, which is tricky to do – so maybe just a long TTL and explicit invalidation from `status` events?

**Comment by grosser** [comment]
we don't have that kind of event ... how about deciding how long to cache depending on if the statuses  are completed (success/failed) ?
(fetch ... fail and retrieve ... decide how long to cache)

**Comment by dasch** [comment]
I guess is the status is pending we just don't cache?

**Comment by grosser** [comment]
maybe cache for a minute or so ... or just don't ... there should be few commits on the release page that are pending ...

**Comment by grosser** [comment]
there is a `github is down` check somewhere, would be good to use that as short-circuit too

**Comment by grosser** [comment]
it's `github_ok?`

**Comment by dasch** [comment]
How about this?

**Comment by dasch** [comment]
@grosser as in, not trying to call the APIs in that case? For the cached entries it should be fine...

**Comment by grosser** [comment]
yes!

File: app/models/github_status.rb [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+
+    cache_key = [self.class.to_s, @repo, @ref]
+
+    @status_response = @cache.read(cache_key)
+
+    # Fetch the data if the cache returned nil.
+    @status_response ||= begin
+      @github.combined_status(@repo, @ref)
+    rescue Octokit::Error
+      nil
```
Comment (grosser) [inline app/models/github_status.rb]: should either cache that nil or return early ?

Comment (dasch) [reply app/models/github_status.rb]: Ah yes.

**Comment by dasch** [comment]
@grosser happy with the `github_ok?` code now?

File: app/models/github_status.rb (hunk 17 above)
Comment (dasch) [reply app/models/github_status.rb]: I've changed the code so nils aren't cached – that should do the trick.

File: app/models/github_status.rb [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    end
+
+    # Don't cache pending statuses, since we expect updates soon.
+    unless @status_response.nil? || @status_response.state == "pending"
```
Comment (grosser) [inline app/models/github_status.rb]: prefer to avoid unless + binary logic
`if @status_response&.state == "pending"`

File: app/helpers/releases_helper.rb (hunk 14 above)
Comment (grosser) [reply app/helpers/releases_helper.rb]: ^

File: app/models/github_status.rb (hunk 13 above)
Comment (dasch) [reply app/models/github_status.rb]: I can give it a try.

**Comment by dasch** [comment]
@grosser one more go?

**Comment by dasch** [comment]
I simplified GithubStatus btw, I think it's much better now that the network code is done procedurally.

File: app/helpers/releases_helper.rb [hunk 19]
```diff
@@ -1,5 +1,19 @@
 # frozen_string_literal: true
 module ReleasesHelper
+  STATUS_ICONS = {
```
Comment (grosser) [inline app/helpers/releases_helper.rb]: also prefix with GITHUB_ ?

File: app/models/github_status.rb [hunk 20]
```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+class GithubStatus
+  Status = Struct.new(:context, :latest_status) do
+    def state
+      latest_status.state
```
Comment (grosser) [inline app/models/github_status.rb]: latest_status->status ?

Comment (grosser) [reply app/models/github_status.rb]: hmm I guess that could get confusing with `state` 🤷‍♂️

File: app/models/github_status.rb [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+      GITHUB.combined_status(repo, ref)
+    rescue Octokit::Error
+      nil
+    end
+
+    # Fall back to a "missing" status.
+    return new("missing", []) if response.nil?
+
+    # Don't cache pending statuses, since we expect updates soon.
+    unless response&.state == "pending"
```
Comment (grosser) [inline app/models/github_status.rb]: don't need the `&.`since nil is handled above

File: test/models/release_test.rb [hunk 22]
```diff
@@ -1,7 +1,7 @@
 # frozen_string_literal: true
 require_relative '../test_helper'
 
-SingleCov.covered! uncovered: 2
+SingleCov.covered! uncovered: 3
```
Comment (grosser) [inline test/models/release_test.rb]: add test for github_status ?

File: app/models/github_status.rb (hunk 20 above)
Comment (dasch) [reply app/models/github_status.rb]: Yeah.

File: app/helpers/releases_helper.rb (hunk 19 above)
Comment (dasch) [reply app/helpers/releases_helper.rb]: :+1:

File: app/models/github_status.rb (hunk 21 above)
Comment (dasch) [reply app/models/github_status.rb]: :+1:

File: test/models/release_test.rb (hunk 22 above)
Comment (dasch) [reply test/models/release_test.rb]: It doesn't contain any logic, and is very unlikely to break. I think a test is superfluous, unless you're adamant about it.

Comment (dasch) [reply test/models/release_test.rb]: The controller test already exercises the code path.

**Comment by dasch** [comment]
@grosser I've included most of your feedback – I'd like to merge this thing, as I can't keep spending time on it. It significantly improves the workflow of release-based projects, so let's get it in.

File: app/views/releases/show.html.erb [hunk 23]
```diff
@@ -16,6 +16,13 @@
   <label>Created</label>
   <%= relative_time(@release.created_at) %>
 
+  <legend>Commit Status</legend>
+  <p>
+    <% @release.github_status.statuses.each do |status| %>
```
Comment (grosser) [inline app/views/releases/show.html.erb]: inside of `github_ok` too ?

**Comment by grosser** [comment]
@ragurney another look ?

File: test/models/release_test.rb (hunk 22 above)
Comment (grosser) [reply test/models/release_test.rb]: hmmm I'd still add the test since we are trying to keep everything fully unit-tested.

**Comment by dasch** [comment]
:heart:

File: app/views/releases/show.html.erb (hunk 23 above)
Comment (dasch) [reply app/views/releases/show.html.erb]: I don't think this page is critical to the workflow... let's see if it becomes a problem.

**Comment by grosser** [comment]
screen shot 2018-09-10 at 11 00 42 am

just saw pending since 6 months ... either a bug or we might want to cache these too after ~1 hour or so
